### PR TITLE
remove query stuff from record page

### PR DESCRIPTION
**Pull request by lievenhey**
rewrite record page to query data asynchronously from RecordHost. This will make implementing ssh much more easier.

### CODE REVIEW

**Comment by lievenhey** [comment]
It shouldn't since we build with kf5 version 98+

**Comment by GitMensch** [comment]
Does the AppImage chrash on your side, too?

**Comment by lievenhey** [comment]
yes, but if I build it from source it works. Still need some time to investigate.

**Comment by GitMensch** [comment]
Only "minor investigation": stacktrace from extracted appimage with the debug symbols manually loaded:

~~~
#/tmp/squashfs-root/usr/bin> gdb -q ./hotspot
Reading symbols from ./hotspot...

warning: Loadable section ".dynstr" outside of ELF segments
  in /tmp/squashfs-root/usr/bin/hotspot
(No debugging symbols found in ./hotspot)
(gdb) symbol-file /tmp/hotspot-debuginfo-v1.4.0-43-g02fd82c/usr/bin/hotspot
Reading symbols from /tmp/hotspot-debuginfo-v1.4.0-43-g02fd82c/usr/bin/hotspot...
(gdb) run
Starting program: /tmp/squashfs-root/usr/bin/hotspot
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/usr/lib64/libthread_db.so.1".
[New Thread 0x7fffdbbc9700 (LWP 3736130)]
QStandardPaths: XDG_RUNTIME_DIR not set, defaulting to '/tmp/runtime-so'
QStandardPaths: XDG_RUNTIME_DIR not set, defaulting to '/tmp/runtime-so'
[New Thread 0x7fffd3fff700 (LWP 3736324)]
[New Thread 0x7fffcb7fe700 (LWP 3736325)]
[New Thread 0x7fffd37fe700 (LWP 3736326)]
[New Thread 0x7fffd2ffd700 (LWP 3736327)]
[New Thread 0x7fffd27fc700 (LWP 3736328)]
[New Thread 0x7fffd1ffb700 (LWP 3736329)]
[New Thread 0x7fffd17fa700 (LWP 3736330)]
[New Thread 0x7fffd0ff9700 (LWP 3736331)]
[New Thread 0x7fffcbfff700 (LWP 3736332)]
[Detaching after fork from child process 3736333]
ASSERT failure in KConfigGroup::readEntry: "accessing an invalid group", file /opt/kde/src/frameworks/kconfig/src/core/kconfiggroup.cpp, line 675

Thread 1 "hotspot" received signal SIGABRT, Aborted.
0x00007ffff3efb37f in raise () from /usr/lib64/libc.so.6
(gdb) bt
#0  0x00007ffff3efb37f in raise () from /usr/lib64/libc.so.6
#1  0x00007ffff3ee5db5 in abort () from /usr/lib64/libc.so.6
#2  0x00007ffff4c53c81 in QMessageLogger::fatal(char const*, ...) const ()
   from /tmp/squashfs-root/usr/bin/../lib/libQt5Core.so.5
#3  0x00007ffff4c5321e in qt_assert_x(char const*, char const*, char const*, int) ()
   from /tmp/squashfs-root/usr/bin/../lib/libQt5Core.so.5
#4  0x00007ffff63f1095 in KConfigGroup::readEntry(char const*, QString const&) const ()
   from /tmp/squashfs-root/usr/bin/../lib/libKF5ConfigCore.so.5
#5  0x000000000048855a in operator() (filePath=..., __closure=0x728c70) at /__w/hotspot/hotspot/src/recordpage.cpp:171
#6  QtPrivate::FunctorCall<QtPrivate::IndexesList<0>, QtPrivate::List<const QString&>, void, RecordPage::RecordPage(QWidget*)::<lambda(const QString&)> >::call (arg=<optimized out>, f=...) at /usr/include/QtCore/qobjectdefs_impl.h:146
#7  QtPrivate::Functor<RecordPage::RecordPage(QWidget*)::<lambda(const QString&)>, 1>::call<QtPrivate::List<QString const&>, void>
    (arg=<optimized out>, f=...) at /usr/include/QtCore/qobjectdefs_impl.h:256
#8  QtPrivate::QFunctorSlotObject<RecordPage::RecordPage(QWidget*)::<lambda(const QString&)>, 1, QtPrivate::List<const QString&>, void>::impl(int, QtPrivate::QSlotObjectBase *, QObject *, void **, bool *) (which=<optimized out>, this_=0x728c60,
    r=<optimized out>, a=<optimized out>, ret=<optimized out>) at /usr/include/QtCore/qobjectdefs_impl.h:443
#9  0x00007ffff4ea316d in ?? () from /tmp/squashfs-root/usr/bin/../lib/libQt5Core.so.5
#10 0x000000000043c355 in RecordHost::clientApplicationChanged (this=this@entry=0x6cc020, _t1=...)
    at /__w/hotspot/hotspot/build/src/hotspot_autogen/EWIEGA46WW/moc_recordhost.cpp:294
#11 0x00000000004c183f in RecordHost::setHost (this=0x6cc020, host=...) at /__w/hotspot/hotspot/src/recordhost.cpp:205
#12 0x000000000048a67a in RecordPage::RecordPage (this=this@entry=0x6cbf70, parent=parent@entry=0x647af0)
    at /__w/hotspot/hotspot/src/recordpage.cpp:215
#13 0x0000000000471467 in MainWindow::MainWindow (this=this@entry=0x647af0, parent=parent@entry=0x0, __in_chrg=<optimized out>,
    __vtt_parm=<optimized out>) at /__w/hotspot/hotspot/src/mainwindow.cpp:104
#14 0x000000000043b187 in main (argc=<optimized out>, argv=<optimized out>) at /__w/hotspot/hotspot/src/main.cpp:195
(gdb) 
~~~

File: src/recordhost.cpp [hunk 1]
```diff
@@ -0,0 +1,342 @@
...
+            emit errorOccurred(tr("Working directory folder is not writable: %1").arg(cwd));
+        } else {
+            emit errorOccurred({});
+            m_cwd = cwd;
+            emit currentWorkingDirectoryChanged(cwd);
+        }
+        return;
+    }
+
+    qWarning() << __FUNCTION__ << "is not implemented for remote";
```
Comment (milianw) [inline src/recordhost.cpp:262]: piping `__FUNCTION__` is useless, just setup `QT_MESSAGE_PATTERN` in your dev env with `%{function}` and you'll get that for free

File: src/recordhost.h [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+        return m_perfCapabilities;
+    }
+
+    RecordType recordType() const
+    {
+        return m_recordType;
+    }
+    void setRecordType(RecordType type);
+
+    void setPids(const QStringList& pids);
```
Comment (milianw) [inline src/recordhost.h:98]: pids of what? document!

File: src/recordhost.h [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+    void isPerfInstalledChanged(bool isInstalled);
+    void outputFileNameChanged(const QString& outputFileName);
+    void recordTypeChanged(RecordType type);
+    void pidsChanged();
+    // ... more
+
+private:
+    bool isLocal() const
+    {
+        return m_host == QStringLiteral("localhost");
```
Comment (milianw) [inline src/recordhost.h:118]: QLatin1String, don't inline in the header

File: src/recordhost.h [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+    }
+
+    QString m_host = {};
```
Comment (milianw) [inline src/recordhost.h:121]: useless `= {}`, remove

File: src/recordhost.h [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+    QString m_error;
+    QString m_cwd;
+    QString m_clientApplication;
+    QString m_outputFileName;
+    bool m_isPerfInstalled = false;
+    PerfCapabilities m_perfCapabilities = {};
```
Comment (milianw) [inline src/recordhost.h:127]: dito

File: src/recordhost.h [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+    JobTracker m_checkPerfCapabilitiesJob;
+    JobTracker m_checkPerfInstalledJob;
+    RecordType m_recordType = LaunchApplication;
+    QStringList m_pids = {};
```
Comment (milianw) [inline src/recordhost.h:131]: dito

File: src/recordhost.h [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+    {
+        return m_host == QStringLiteral("localhost");
+    }
+
+    QString m_host = {};
+    QString m_error;
+    QString m_cwd;
+    QString m_clientApplication;
+    QString m_outputFileName;
+    bool m_isPerfInstalled = false;
```
Comment (milianw) [inline src/recordhost.h:126]: pack below `m_recordType` to reduce padding

File: src/recordpage.h [hunk 8]
```diff
@@ -14,6 +14,9 @@
 
 #include "processlist.h"
 
+// TODO fix RecordType
```
Comment (milianw) [inline src/recordpage.h:17]: yes do that then, or remove this comment

File: src/recordpage.h [hunk 9]
```diff
@@ -26,20 +29,12 @@ class ProcessModel;
 class ProcessFilterModel;
 class MultiConfigWidget;
 class PerfOutputWidget;
+class RecordHost;
```
Comment (milianw) [inline src/recordpage.h:32]: you have an include for that too here? either/or, ideally only the forward declaration

File: src/settingsdialog.cpp [hunk 10]
```diff
@@ -32,17 +33,24 @@ KConfigGroup config()
 {
     return KSharedConfig::openConfig()->group("PerfPaths");
 }
+
+QIcon icon()
+{
+    return QIcon::fromTheme(QStringLiteral("preferences-system-windows-behavior"));
```
Comment (milianw) [inline src/settingsdialog.cpp:39]: static const auto icon = ...;
return icon;

File: src/util.cpp [hunk 11]
```diff
@@ -435,3 +435,18 @@ KParts::ReadOnlyPart* Util::createPart(const QString& pluginName)
...
 }
+
+QString Util::sudoUtil()
+{
+    const auto commands = {
+        QStringLiteral("pkexec"), QStringLiteral("kdesudo"), QStringLiteral("kdesu"),
+        // gksudo / gksu seem to close stdin and thus the elevate script doesn't wait on read
+    };
+    for (const auto& cmd : commands) {
+        QString util = QStandardPaths::findExecutable(cmd);
```
Comment (milianw) [inline src/util.cpp:446]: auto

File: tests/integrationtests/tst_perfparser.cpp [hunk 12]
```diff
@@ -159,9 +160,21 @@ class TestPerfParser : public QObject
 private slots:
     void initTestCase()
     {
-        if (!PerfRecord::isPerfInstalled()) {
+        RecordHost host;
+        host.setHost(QStringLiteral("localhost"));
```
Comment (milianw) [inline tests/integrationtests/tst_perfparser.cpp:164]: default to localhost in the ctor?

File: tests/integrationtests/tst_perfparser.cpp [hunk 13]
```diff
@@ -159,9 +160,21 @@ class TestPerfParser : public QObject
...
+        QSignalSpy capabilitiesSpy(&host, &RecordHost::perfCapabilitiesChanged);
+        QSignalSpy installedSpy(&host, &RecordHost::isPerfInstalledChanged);
+        Q_ASSERT(installedSpy.wait());
```
Comment (milianw) [inline tests/integrationtests/tst_perfparser.cpp:167]: QVERIFY

File: tests/integrationtests/tst_perfparser.cpp [hunk 14]
```diff
@@ -159,9 +160,21 @@ class TestPerfParser : public QObject
...
+        if (!host.isPerfInstalled()) {
             QSKIP("perf is not available, cannot run integration tests.");
         }
+
+        if (capabilitiesSpy.count() == 0) {
+            Q_ASSERT(capabilitiesSpy.wait());
```
Comment (milianw) [inline tests/integrationtests/tst_perfparser.cpp:173]: dito

File: tests/integrationtests/tst_perfparser.cpp [hunk 15]
```diff
@@ -159,9 +160,21 @@ class TestPerfParser : public QObject
...
+        }
+        m_capabilities = host.perfCapabilities();
+
+        qDebug() << m_capabilities.canCompress << m_capabilities.canUseAio;
```
Comment (milianw) [inline tests/integrationtests/tst_perfparser.cpp:177]: qInfo, if you always want to show this

File: src/recordhost.h [hunk 16]
```diff
@@ -0,0 +1,134 @@
...
+enum RecordType
+{
+    LaunchApplication = 0,
+    AttachToProcess,
+    ProfileSystem,
+    NUM_RECORD_TYPES
+};
+Q_DECLARE_METATYPE(RecordType)
+
+// all of recordpage
```
[the remark this answers is not in the document]
Comment (lievenhey) [reply src/recordhost.h:22]: A note I forgot to remove

File: tests/integrationtests/tst_perfparser.cpp (hunk 15 above)
Comment (lievenhey) [reply tests/integrationtests/tst_perfparser.cpp:177]: no I just forgot to remove printf debugging

File: src/recordhost.h (hunk 3 above)
Comment (lievenhey) [reply src/recordhost.h]: Clazy doesn't like QLatin1String 😛

**Comment by GitMensch** [comment]
@lievenhey Can you please give a an overview where this PR stands? I guess #359 waits for this to be handled first.

**Comment by lievenhey** [comment]
I am currently waiting for the new elevate system. Once that is merged, I will tackle this and then ssh

**Comment by GitMensch** [comment]
Thank you, so #493 comes before this and this becomes before ssh - thank you for the clarification.

**Comment by GitMensch** [comment]
@milianw: seems like all of your review comments are handled, so possibly ready for pull?

**Comment by GitMensch** [comment]
.. but docs (screenshots) should likely be adjusted either here or with a follow-up PR.

File: src/recordhost.cpp [hunk 17]
```diff
@@ -0,0 +1,349 @@
...
+    QObject::connect(&process, &QProcess::errorOccurred, &process, reportError);
+    process.start(perfPath, arguments);
+    if (!process.waitForFinished(1000) || process.exitCode() != 0)
+        reportError();
+    return process.readAllStandardOutput();
+}
+
+QByteArray perfRecordHelp(const QString& perfPath)
+{
+    QByteArray recordHelp = [&perfPath]() {
```
Comment (milianw) [inline src/recordhost.cpp:53]: simplify, the outer lambda/IILE isn't needed anymore, now that we don't store the `recordHelp` in a static anymore

File: src/util.cpp [hunk 18]
```diff
@@ -435,3 +435,18 @@ KParts::ReadOnlyPart* Util::createPart(const QString& pluginName)
     return service->createInstance<KParts::ReadOnlyPart>();
 #endif
 }
+
+QString Util::sudoUtil()
```
Comment (milianw) [inline src/util.cpp:439]: not needed anymore?

File: src/recordhost.cpp [hunk 19]
```diff
@@ -11,6 +11,7 @@
 #include <QProcess>
 #include <QStandardPaths>
 #include <QThread>
+#include <qstringliteral.h>
```
Comment (milianw) [inline src/recordhost.cpp:14]: nope, shouldn't be needed (or include QString instead)

**Review by milianw — CHANGES_REQUESTED** [review]

**Comment by milianw** [comment]
> .. but docs (screenshots) should likely be adjusted either here or with a follow-up PR.

nothing user-visible is changed in this change, it's purely internal

... except that we now can configure the path to the perf that should be used - but that's so minor, we don't need to have this imo :) that said, patches welcome obviously if you want to document this!

**Pull request merged**